reactor-netty-http/src/main/java/reactor/netty/http/client/Http2Pool.java

```diff
@@ -0,0 +1,820 @@
+			actual.onSubscribe(borrower);
+		}
+	}
+
+	static final class Http2PooledRef extends AtomicBoolean implements PooledRef<Connection>, PooledRefMetadata {
+
+		final int acquireCount;
+		final Slot slot;
+
+		Http2PooledRef(Slot slot) {
+			this.acquireCount = 0;
+			this.slot = slot;
+		}
+
+		@Override
+		public int acquireCount() {
+			return 1;
+		}
+
+		@Override
+		public long allocationTimestamp() {
+			return 0;
+		}
+
+		@Override
+		public long idleTime() {
+			return 0;
+		}
+
+		@Override
+		public Mono<Void> invalidate() {
+			return Mono.defer(() -> {
+				if (compareAndSet(false, true)) {
+					ACQUIRED.decrementAndGet(slot.pool);
+					return slot.pool.destroyPoolable(this).doFinally(st -> slot.pool.drain());
+				}
+				else {
+					return Mono.empty();
+				}
+			});
+		}
+
+		@Override
+		public long lifeTime() {
+			return 0;
+		}
+
+		@Override
+		public PooledRefMetadata metadata() {
+			return this;
+		}
+
+		@Override
+		public Connection poolable() {
+			return slot.connection;
+		}
+
+		@Override
+		public Mono<Void> release() {
+			return invalidate();
+		}
+
+		@Override
+		public long releaseTimestamp() {
+			return 0;
+		}
+
+		@Override
+		public String toString() {
+			return "PooledRef{poolable=" + slot.connection + '}';
+		}
+	}
+
+	static final class Slot {
+
+		volatile int concurrency;
+		static final AtomicIntegerFieldUpdater<Slot> CONCURRENCY =
+				AtomicIntegerFieldUpdater.newUpdater(Slot.class, "concurrency");
+
+		volatile int state;
+		static final AtomicIntegerFieldUpdater<Slot> STATE = AtomicIntegerFieldUpdater.newUpdater(Slot.class, "state");
+
+		static final int STATE_DEACTIVATED = 0;
+		static final int STATE_ACTIVATED = 1;
+		static final int STATE_INVALIDATED = 2;
+
+		final Connection connection;
+		final long creationTimestamp;
+		final Http2Pool pool;
+
+		Slot(Http2Pool pool, Connection connection) {
+			this.connection = connection;
+			this.creationTimestamp = pool.clock.millis();
+			this.pool = pool;
+			this.state = STATE_DEACTIVATED;
+		}
+
+		boolean activate() {
+			if (STATE.compareAndSet(this, STATE_DEACTIVATED, STATE_ACTIVATED)) {
+				if (log.isDebugEnabled()) {
+					log.debug(format(connection.channel(), "Channel activated"));
+				}
+				return true;
+			}
+			return false;
+		}
+
+		boolean canOpenStream() {
+			Http2FrameCodec frameCodec = connection.channel().pipeline().get(Http2FrameCodec.class);
+			if (frameCodec != null) {
+				int maxActiveStreams = frameCodec.connection().local().maxActiveStreams();
+				int concurrency = this.concurrency;
+				return concurrency < maxActiveStreams;
+			}
+			return false;
+		}
+
+		int concurrency() {
+			return concurrency;
+		}
+
+		void deactivate() {
+			if (STATE.compareAndSet(this, STATE_ACTIVATED, STATE_DEACTIVATED)) {
+				if (log.isDebugEnabled()) {
+					log.debug(format(connection.channel(), "Channel deactivated"));
+				}
+				offerSlot(this);
+			}
+		}
+
+		void decrementConcurrency() {
```

Comment:
I'm still a bit worried about these methods, and the fact that there is a separate `STATE` and `CONCURRENCY`... Here the `incrementConcurrency()`+`decrementConcurrency()`+`concurrency()` break the atomic benefit of `CONCURRENCY`. Can the STATE and CONCURRENCY be collapsed into a single int and the methods reworked to use the CAS / incrementAndGet / decrementAndGet better ?